### PR TITLE
Use procedural macro to generate static public keys

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
Lazy static initialization of sysvar ids does not work in BPF programs

#### Summary of Changes
Create a procedural macro which will check program id validity at compile time

Fixes https://github.com/solana-labs/solana/issues/7218

### CODE REVIEW

File: sdk/macro/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        if !input.is_empty() {
+            let stream: proc_macro2::TokenStream = input.parse()?;
+            return Err(syn::Error::new_spanned(stream, "unexpected token"));
+        }
+
+        let id_vec = bs58::decode(id_literal.value())
+            .into_vec()
+            .map_err(|_| syn::Error::new_spanned(&id_literal, "failed to decode base58 id"))?;
+        let id_array = <[u8; 32]>::try_from(<&[u8]>::clone(&&id_vec[..]))
+            .map_err(|_| syn::Error::new_spanned(&id_literal, "id is not 32 bytes long"))?;
```
Comment (jstarry) [inline sdk/macro/src/lib.rs]: Minimal copy/paste needed to avoid cyclical dependency between solana-sdk and solana-sdk-macro

**Comment by jstarry** [comment]
@jackcmay I confirmed that this fixes the message feed issue, no more `.bss` section with this change

**Comment by mvines** [comment]
Can we add an explicit BPF test for this please.  I bet it sucked to trace this problem down from the downstream example  app randomly started silently failing.

File: sdk/macro/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
+//! Convenience macro to declare a static public key and functions to interact with it
+//!
+//! bs58_string: bs58 string representation the program's id
```
Comment (jackcmay) [inline sdk/macro/src/lib.rs]: nit: "representation **of** the program's id"

File: programs/move_loader/src/lib.rs [hunk 3]
```diff
@@ -4,10 +4,9 @@ pub mod error_mappers;
 pub mod processor;
 
 use crate::processor::process_instruction;
-use solana_sdk::move_loader;
 
 solana_sdk::declare_program!(
-    move_loader::BS58_STRING,
+    "MoveLdr111111111111111111111111111111111111",
```
Comment (jackcmay) [inline programs/move_loader/src/lib.rs]: Can we avoid duplicating the definition of the id string between here and the sdk?

**Comment by jackcmay** [comment]
Awesome use of procedural macros!

File: programs/move_loader/src/lib.rs (hunk 3 above)
Comment (jstarry) [reply programs/move_loader/src/lib.rs]: Yup, done

File: sdk/src/instruction_processor_utils.rs [hunk 4]
```diff
@@ -51,17 +54,70 @@ macro_rules! solana_entrypoint(
...
+    );
+    ($program_id:path, $name:ident, $entrypoint:expr) => (
+        declare_program!(@expand $program_id, $name, $entrypoint);
+    );
     ($bs58_string:expr, $name:ident, $entrypoint:expr) => (
         $crate::declare_id!($bs58_string);
+        declare_program!(@declare $name, $entrypoint);
+    );
+    (@expand $program_id:expr, $name:ident, $entrypoint:expr) => (
+        pub static ID: ::solana_sdk::pubkey::Pubkey = $program_id;
```
Comment (jackcmay) [inline sdk/src/instruction_processor_utils.rs]: Why not usee `$crate` here for `::solana_sdk'?

Comment (jstarry) [reply sdk/src/instruction_processor_utils.rs]: Ah thanks, my mistake. I missed updating that one after copying from the new proc macro

`$crate` is used in declarative macros and absolute paths are used in procedural macros

File: sdk/src/instruction_processor_utils.rs [hunk 5]
```diff
@@ -51,17 +54,70 @@ macro_rules! solana_entrypoint(
...
+        $crate::declare_program!(@expand $program_id, $name, $entrypoint);
+    );
     ($bs58_string:expr, $name:ident, $entrypoint:expr) => (
         $crate::declare_id!($bs58_string);
+        $crate::declare_program!(@declare $name, $entrypoint);
+    );
+    (@expand $program_id:expr, $name:ident, $entrypoint:expr) => (
+        pub static ID: $crate::pubkey::Pubkey = $program_id;
+
+        pub fn check_id(id: &$crate::pubkey::Pubkey) -> bool {
```
Comment (jackcmay) [inline sdk/src/instruction_processor_utils.rs]: Be nice to get rid of this duplicate code and its a little weird having it in `declare_program`.  Not a blocker

Comment (jstarry) [reply sdk/src/instruction_processor_utils.rs]: Yeah, good call. Fixed

**Pull request merged**